### PR TITLE
Recipe: Reinforcement learning for generative models using MindSpeed-mm as the backend (DanceGRPO)

### PR DESCRIPTION
**Pull request by ruanhao566**
Summary
------
This PR adding a recipe to the verl-recipe repository that demonstrates how to integrate the verl reinforcement learning framework with [MindSpeed MM](https://gitcode.com/Ascend/MindSpeed-MM/blob/master/README.md) (Ascend Multi-modal Model Suite) to implement RLHF (Reinforcement Learning from Human Feedback) training for generative models on NPU (Ascend AI Processors). The recipe will feature the [DanceGRPO](https://dancegrpo.github.io/) algorithm, using [Wan2.2-5B](https://github.com/Wan-Video/Wan2.2) as the policy model and [HPSv3](https://github.com/MizzenAI/HPSv3) as the reward model, providing an end-to-end training workflow.

Result
------
Using the WAN2.2 model to train for more than 100 steps, the average score is improved by 30%.
image

### CODE REVIEW

File: dance_grpo/dance_grpo_mindspeed_mm/main_dance.py [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+        return True
+    return False
+
+
+if __name__ == "__main__":
+    current_dir = os.path.dirname(os.path.abspath(__file__))
+    # TODO temporary repair
+    file_to_modify = f"{current_dir}/../../mindspeed_mm/models/common/embeddings/pos_embeddings.py"
+    remove_autocast_from_file(file_to_modify)
+    main()
```
[the remark this answers is not in the document]
Comment (ruanhao566) [reply dance_grpo/dance_grpo_mindspeed_mm/main_dance.py:307]: This issue does not affect functions. The version has been provided to the test personnel, and the issue will be resolved together with the new model.

File: dance_grpo/dance_grpo_mindspeed_mm/rollout.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        get_torch_device().empty_cache()
+        return batch
+
+    @torch.no_grad()
+    def generate_sequences(self, prompts: DataProto) -> DataProto:
+        file_index = prompts.non_tensor_batch['index']
+        batch_size = len(file_index)
+        batch_prompts = prompts.chunk(chunks=batch_size)
+        output = [self._generate_minibatch(p[0], index) for index, p in enumerate(batch_prompts)]
+        output = DataProto.concat(output)
```
[the remark this answers is not in the document]
Comment (ruanhao566) [reply dance_grpo/dance_grpo_mindspeed_mm/rollout.py:105]: This issue does not affect functions. The version has been provided to the test personnel, and the issue will be resolved together with the new model.

File: dance_grpo/dance_grpo_mindspeed_mm/diffusion_workers.py [hunk 3]
```diff
@@ -0,0 +1,528 @@
...
+        self._is_actor = role in ["actor", "hybrid", "actor_rollout"]
+        self._is_rollout = role in ["rollout", "hybrid"]
+        # Init mindspeed args
+        import mindspeed.args_utils as args_utils
+        with open(f'{self.config.actor.model_args_path}/mindspeed_args.pkl', 'rb') as f:
+            args_utils._MINDSPEED_ARGS = pickle.load(f)
+        # TODO Mindspeed patch init
+        import mindspeed.megatron_adaptor
+        with open(f'{self.config.actor.model_args_path}/mm_args.pkl', 'rb') as f:
+            args = pickle.load(f)
```
[the remark this answers is not in the document]
Comment (ruanhao566) [reply dance_grpo/dance_grpo_mindspeed_mm/diffusion_workers.py]: This issue does not affect functions. The version has been provided to the test personnel, and the issue will be resolved together with the new model.

File: dance_grpo/dance_grpo_mindspeed_mm/DESIGN_CN.md [hunk 4]
```diff
@@ -0,0 +1,596 @@
...
+│                         Step 3: GRPO Update                         │
+│  ┌───────────────────────────────────────────────────────────────┐  │
+│  │  1. Compute group advantages                                  │  │
+│  │     - Group normalization: (r - mean) / std                   │  │
+│  │  2. Sample timesteps for training                             │  │
+│  │  3. For each timestep:                                        │  │
+│  │     a. Forward pass (compute new log_probs)                   │  │
+│  │     b. Compute ratio = exp(new_log - old_log)                 │  │
+│  │     c. PPO clipping on ratio                                  │  │
+│  │     d. Compute policy loss                                    │  │
```
[the remark this answers is not in the document]
Comment (ruanhao566) [reply dance_grpo/dance_grpo_mindspeed_mm/DESIGN_CN.md:91]: This issue does not affect functions. The version has been provided to the test personnel, and the issue will be resolved together with the new model.

File: dance_grpo/dance_grpo_mindspeed_mm/config/optimizer.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    """
+    import importlib
+
+    optimizer_args = {
+        "lr": config.lr,
+        "weight_decay": config.weight_decay,
+    }
+
+    optimizer_name_lower = config.optimizer.lower()
+    if "adam" in optimizer_name_lower or "ademamix" in optimizer_name_lower:
```
[the remark this answers is not in the document]
Comment (ruanhao566) [reply dance_grpo/dance_grpo_mindspeed_mm/config/optimizer.py:79]: This issue does not affect functions. The version has been provided to the test personnel, and the issue will be resolved together with the new model.

File: dance_grpo/dance_grpo_mindspeed_mm/dataset/data_process.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        print(f"Conversion failed: {str(e)}")
+
+
+if __name__ == "__main__":
+    current_dir = os.path.dirname(os.path.abspath(__file__))
+    print(f"Current_dir: {current_dir}")
+    data_text = current_dir + '/../data/prompt.txt'
+    data_json = current_dir + '/../data/prompt.json'
+    txt_to_json(data_text, data_json)
+    data_save = current_dir + '/../data/parquet/'
```
[the remark this answers is not in the document]
Comment (ruanhao566) [reply dance_grpo/dance_grpo_mindspeed_mm/dataset/data_process.py]: This issue does not affect functions. The version has been provided to the test personnel, and the issue will be resolved together with the new model.

File: dance_grpo/dance_grpo_mindspeed_mm/diffusion_workers.py [hunk 7]
```diff
@@ -0,0 +1,528 @@
...
+        with open(file_path, "a", encoding="utf-8") as f:
+            import json
+            f.write(json.dumps(context, ensure_ascii=False) + "\n")
+
+    def save_rollout_result(self, hps_score, prompt, images_path, global_steps):
+        save_path = self.config.rollout.result.save.path
+        if save_path is None:
+            raise ValueError('the rollout result save path is None')
+        os.makedirs(save_path, exist_ok=True)
+        import shutil
```
[the remark this answers is not in the document]
Comment (ruanhao566) [reply dance_grpo/dance_grpo_mindspeed_mm/diffusion_workers.py:527]: This issue does not affect functions. The version has been provided to the test personnel, and the issue will be resolved together with the new model.

File: dance_grpo/dance_grpo_mindspeed_mm/diffusion_workers.py [hunk 8]
```diff
@@ -0,0 +1,528 @@
...
+        for i in range(latents.shape[0]):
+            images_path = images[i]
+            prompt = prompts[i][0]['content']
+            global_steps = data.non_tensor_batch["global_steps"][i]
+            with torch.no_grad():
+                image = video_first_frame_to_pil(images_path)
+                base, _ = os.path.splitext(images_path)
+                png_path = base + '.png'
+                image.save(png_path, format="PNG")
+                hps_score = self.reward_module.reward([png_path], [prompt])
```
[the remark this answers is not in the document]
Comment (ruanhao566) [reply dance_grpo/dance_grpo_mindspeed_mm/diffusion_workers.py:271]: This issue does not affect functions. The version has been provided to the test personnel, and the issue will be resolved together with the new model.

File: dance_grpo/dance_grpo_mindspeed_mm/README.md [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+
+Wan2.2 5B model download address:
+
+https://huggingface.co/Wan-AI/Wan2.2-TI2V-5B-Diffusers
+
+The downloaded model is in the huggingface format and needs to be converted to the dcp format for training. For details, see the following section. If you need to convert the model back to the huggingface format after training, see the section about converting the dcp format to hf.
+
+### 3.1 HF to DCP ###
+
+1.  Weight of the downloaded Wan2.2 model`transformer`In the mm root directory, run the following script to convert the weight:
```
[the remark this answers is not in the document]
Comment (ruanhao566) [reply dance_grpo/dance_grpo_mindspeed_mm/README.md:103]: This issue does not affect functions. The version has been provided to the test personnel, and the issue will be resolved together with the new model.

File: dance_grpo/dance_grpo_mindspeed_mm/actor.py [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+                text_hidden_states,
+                negative_text_hidden_states,
+            )
+        return log_probs
+
+    def compute_log_prob(self, data: DataProto) -> torch.Tensor:
+        raise NotImplementedError("compute_log_prob is not implemented for DataParallelPPOActor")
+
+    def update_policy(self, data: DataProto) -> dict:
+        raise NotImplementedError("update_policy is not implemented for DataParallelPPOActor")
```
[the remark this answers is not in the document]
Comment (ruanhao566) [reply dance_grpo/dance_grpo_mindspeed_mm/actor.py:95]: This issue does not affect functions. The version has been provided to the test personnel, and the issue will be resolved together with the new model.

File: mm_dance_grpo/DESIGN.md [hunk 11]
```diff
@@ -0,0 +1,609 @@
...
+    
+    Process:
+    1. Get latents, old_log_probs, rewards.
+    2. Compute advantages.
+    3. Sample training timesteps.
+    4. For each timestep:
+       a. Forward pass to compute new_log_probs.
+       b. Compute probability ratio ratio = exp(new_log - old_log).
+       c. PPO clipping: clipped_ratio = clamp(ratio, 1-ε, 1+ε).
+       d. Compute loss: loss = -adv * max(ratio, clipped_ratio).
```
[the remark this answers is not in the document]
Comment (ruanhao566) [reply mm_dance_grpo/DESIGN.md]: This issue does not affect functions. The version has been provided to the test personnel, and the issue will be resolved together with the new model.
